Sources/JWT/JWKSCache.swift

```diff
@@ -3,84 +3,137 @@ import Vapor
+        logger.debug("Requesting JWKS from \(self.uri).")
+        let keys = self.requestKeys(logger: logger)
+        self.currentRequest = keys
 
-                    return self.eventLoop.makeSucceededFuture(decoded)
-            }.hop(to: req.eventLoop)
+        // Once the key request finishes, clear the current
+        // request and return the keys.
+        return keys.map { keys in
+            self.currentRequest = nil
+            return keys
         }
     }
 
-    private func update(expires: Date?, etag: String?, jwks: JWKS) {
-        if let expires = expires {
-            self.jwks = jwks
-            self.etag = etag
-            self.cacheUntil = expires
-        } else {
-            self.jwks = nil
-            self.etag = nil
-            self.cacheUntil = nil
+    private func requestKeys(logger: Logger) -> EventLoopFuture<JWKS> {
+        // Add cached eTag header to this request if we have it.
+        var headers: HTTPHeaders = [:]
+        if let eTag = self.cachedJWKS?.eTag {
+            headers.add(name: .ifNoneMatch, value: eTag)
+        }
+
+        // Store the requested-at date to calculate expiration date.
+        let requestSentAt = Date()
+
+        // Send the GET request for the JWKs.
+        return self.client.get(
+            self.uri, headers: headers
+        ).flatMapThrowing { response in
+            // Synchronize access to shared state.
+            self.sync.lock()
+            defer { self.sync.unlock() }
+
+            let expirationDate = response.headers.expirationDate(requestSentAt: requestSentAt)
+            let eTag = response.headers.firstValue(name: .eTag)
+            switch response.status {
+            case .notModified:
+                // The cached JWKS are still the latest version.
+                logger.debug("Cached JWKS are still valid.")
+                guard var cachedJWKS = self.cachedJWKS else {
+                    throw Error.missingCache
+                }
+
+                // Update the JWKS cache if there is an expiration date.
+                if let expirationDate = expirationDate {
+                    // Update the cache metadata.
+                    cachedJWKS.cacheUntil = expirationDate
+                    cachedJWKS.eTag = eTag
+                    self.cachedJWKS = cachedJWKS
+                } else {
+                    self.cachedJWKS = nil
+                }
+                return cachedJWKS.jwks
+            case .ok:
+                // New JWKS have been returned.
+                logger.debug("New JWKS have been returned.")
+                let jwks = try response.content.decode(JWKS.self)
+
+                // Cache the JWKS if there is an expiration date.
+                if let expirationDate = expirationDate {
+                    if var cachedJWKS = self.cachedJWKS {
+                        // Update the existing cache.
+                        cachedJWKS.cacheUntil = expirationDate
+                        cachedJWKS.eTag = eTag
```

Comment:
etag and cache are two separate things.  Should set the etag if it exists, even if there is no caching.